src/topology_sampler.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019-2021 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+
+#include "topology_sampler.hpp"
+#include "numerical_utils.hpp"
+#include "unrooted_sbn_instance.hpp"
+#include "rooted_sbn_instance.hpp"
+#include "subsplit_dag.hpp"
+
+Node::NodePtr TopologySampler::SampleTopology(const Input &input,
+                                              bool is_rooted) const {
+  // Start by sampling a rootsplit.
+  size_t rootsplit_index =
+      SampleIndex(input, std::pair<size_t, size_t>(0, input.RootsplitCount()));
+  const Bitset &rootsplit = input.RootsplitsAt(rootsplit_index);
+  auto topology = SampleTopology(input, rootsplit);
+  if (!is_rooted) {
+    topology = topology->Deroot();
+  }
+  topology->Polish();
+  return topology;
+}
+
+// Sample an integer index in [range.first, range.second) according to
+// SBNParameters.
+size_t TopologySampler::SampleIndex(const Input &input, Range range) const {
+  const auto &[start, end] = range;
+  Assert(start < end && static_cast<Eigen::Index>(end) <= input.SBNParameters().size(),
+         "SampleIndex given an invalid range.");
+  // We do not want to overwrite sbn_parameters so we make a copy.
+  EigenVectorXd sbn_parameters_subrange =
+      input.SBNParameters().segment(start, end - start);
+  NumericalUtils::ProbabilityNormalizeInLog(sbn_parameters_subrange);
+  NumericalUtils::Exponentiate(sbn_parameters_subrange);
+  std::discrete_distribution<> distribution(sbn_parameters_subrange.begin(),
+                                            sbn_parameters_subrange.end());
+  // We have to add on start because we have taken a slice of the full
+  // array, and the sampler treats the beginning of this slice as zero.
+  auto result =
+      start + static_cast<size_t>(distribution(mersenne_twister_.GetGenerator()));
+  Assert(result < end, "SampleIndex sampled a value out of range.");
+  return result;
+}
+
+// The input to this function is a parent subsplit (of length 2n).
+Node::NodePtr TopologySampler::SampleTopology(const Input &input,
+                                              const Bitset &parent_subsplit) const {
+  auto process_subsplit = [this, &input](const Bitset &parent) {
+    auto singleton_option = parent.SubsplitGetClade(1).SingletonOption();
+    if (singleton_option) {
+      return Node::Leaf(*singleton_option);
+    }
+    auto child_index = SampleIndex(input, input.ParentToRangeAt(parent));
+    return SampleTopology(input, input.IndexToChildAt(child_index));
```

Comment:
Using this linear time function means that we the whole topology sampling is quadratic time when it should be linear. 

Let's think a little more about exactly how we want to set up this topology sampling-- I think that we might need a subsplit DAG-specific method rather than doing something generic here. 

The range of indices that we want should be available as the leafward edges from the node, right? See the documentation in `subsplit_dag.hpp`: 

```
// - The nodes of the DAG have vectors of indices representing their edges to other
// nodes. These include edges to the children (leafward edges) and also the edges that
// are connecting to that node (rootward edges). Furthermore, these sets of edges are
// separated into two classes: those that split apart the left component of the subsplit
// ("rotated" edges) and those that split apart the right component of the subsplit
// ("sorted" edges). The nodes of the DAG also include bitsets describing the subsplit
// that they represent.
```